File: backend/utils/logger.js
function _format(level, msg, meta) {
  const out = {
    ts: new Date().toISOString(),
    level,
    msg,
    ...meta
  };
  return JSON.stringify(out);
}

function info(msg, meta = {}) {
  console.log(_format('info', msg, meta));
}

function warn(msg, meta = {}) {
  console.warn(_format('warn', msg, meta));
}

function error(msg, meta = {}) {
  console.error(_format('error', msg, meta));
}

module.exports = { info, warn, error };

File: bioservice/utils/logger.py
import json
from datetime import datetime
import sys

def _emit(level: str, msg: str, **meta):
    out = {
        "ts": datetime.utcnow().isoformat() + 'Z',
        "level": level,
        "msg": msg,
        **meta
    }
    print(json.dumps(out), file=sys.stdout)

def info(msg: str, **meta):
    _emit('info', msg, **meta)

def warn(msg: str, **meta):
    _emit('warn', msg, **meta)

def error(msg: str, **meta):
    _emit('error', msg, **meta)
